(feat) O3-5255: Add optional discharge note to ward patient discharge workspace

sourav-jyoti opened the pull request:
## Requirements

- [x] This PR has a title that briefly describes the work done including the ticket number. If there is a ticket, make sure your PR title includes a [conventional commit](https://o3-docs.openmrs.org/docs/frontend-modules/contributing.en-US#contributing-guidelines) label. See existing PR titles for inspiration.
- [x] My work is based on designs, which are linked or shown either in the Jira ticket or the description below.
- [x] My work includes tests or is validated by existing tests.

## Summary

Adds an optional free-text note field to the ward patient discharge workspace, allowing clinicians to record a reason or context when discharging a patient. The note is saved as an observation using the `consultFreeTextCommentsConcept` on the exit-from-inpatient encounter.

Other changes:
- Rewrites the discharge handler from promise chains to async/await
- Adds Cancel and Confirm buttons (replacing the single ghost button)
- Wires up `hasUnsavedChanges` so closing the workspace with a draft note prompts for confirmation before discarding
- Improves error handling with a structured fallback chain
- Adds test coverage for the discharge workspace

## Screenshots

### Before

https://github.com/user-attachments/assets/adb8da7f-a9d3-4cf8-b8ca-3a45a4a8ab06

### After

https://github.com/user-attachments/assets/8246c1af-2644-4976-9ae3-94f019bc0d73

## Related Issue

https://openmrs.atlassian.net/issues/O3-5255

## Other

sourav-jyoti committed:
fix/O3-5255

sourav-jyoti committed:
feat/O3-5255 2

sourav-jyoti commented on packages/esm-ward-app/src/ward-workspace/patient-discharge/patient-discharge.workspace.tsx:
```diff
@@ -1,13 +1,35 @@
 import styles from './patient-discharge.scss';
 
+const createNoteSchema = (translate: (key: string, defaultMessage: string) => string) =>
+  z.object({
```
i have applied this check currently on the frontend as mentioned by anish_dhodari in the [ticket ](https://openmrs.atlassian.net/issues/?filter=-1&selectedIssue=O3-5255) " A **mandatory discharge** note field to record the reason for discharge."

so should i remove it completely

sourav-jyoti committed:
feat/O3-5255 3

chibongho commented on packages/esm-ward-app/src/ward-workspace/patient-discharge/patient-discharge.workspace.tsx:
```diff
@@ -18,41 +40,80 @@ export default function PatientDischargeWorkspace({
+          }
+        })
+        .then((response) => {
+          if (!response?.ok) {
```
I know the old code had this `response.ok` check, but I don't think we need it.

https://o3-docs.openmrs.org/docs/coding-conventions/error-handling.en-US

sourav-jyoti committed:
feat/O3-5255

sourav-jyoti committed:
feat/O3-5255

sourav-jyoti committed:
fix

sourav-jyoti commented:
@denniskigen i have updated the according to the new requirements 
1. removed the confirmation modal
2. made the note optional
3. used modern try catch block as suggested by chibongho

### Updated screenrecording

https://github.com/user-attachments/assets/9378abf4-c324-4b0a-8331-be0221594cfc

sourav-jyoti committed:
fix

denniskigen committed:
Merge branch 'main' into feat/O3-5255

denniskigen reviewed (commented):
Thanks for iterating on this, @sourav-jyoti — the async/await conversion is cleaner, and the test coverage is thorough. A few suggestions below.

One thing to update before merge: the PR description still references the confirmation modal and mandatory note field, but the current code has neither (per earlier reviewer feedback). Could you update the description to match the current implementation?

sourav-jyoti committed:
update

sourav-jyoti commented:
@denniskigen could you pls review it

sourav-jyoti committed:
Merge branch 'main' into feat/O3-5255

sourav-jyoti committed:
update

sourav-jyoti committed:
update

anish-mekom commented:
Hi @denniskigen, requesting for your review on the updates.

denniskigen committed:
Merge branch 'main' into feat/O3-5255

denniskigen commented on packages/esm-ward-app/src/ward-workspace/patient-discharge/patient-discharge.workspace.tsx:
```diff
@@ -66,10 +89,10 @@ export default function PatientDischargeWorkspace({
         <div className={styles.patientWorkspaceBanner}>
```
This workspace now has editable state (`dischargeNote`), but `Workspace2` doesn't receive `hasUnsavedChanges`. If a user types a note and closes the workspace, the draft is silently lost. Other ward workspaces wire this up, e.g. `cancel-admission-request.component.tsx` and `notes.workspace.tsx`. Please pass `hasUnsavedChanges={dischargeNote.length > 0}`.

denniskigen commented on packages/esm-ward-app/src/ward-workspace/patient-discharge/patient-discharge.workspace.tsx:
```diff
@@ -14,47 +20,64 @@ export default function PatientDischargeWorkspace({
+
+        showSnackbar({
+          title: t('patientWasDischarged', 'Patient was discharged'),
+          kind: 'success',
```
In the original code, `mutate()` was called in `finally`, so it ran on both success and failure. Moving it into the `try` block means a partial failure (encounter created but bed removal throws) leaves the ward view stale. Consider moving it back to `finally`.

denniskigen commented on packages/esm-ward-app/src/ward-workspace/patient-discharge/patient-discharge.workspace.tsx:
```diff
@@ -83,20 +106,37 @@ export default function PatientDischargeWorkspace({
+              />
+            </ResponsiveWrapper>
+          </div>
           <ButtonSet className={styles.buttonSet}>
```
Two things: "discharging" should be sentence case ("Discharging"), and concatenating the ellipsis is fragile for i18n. Consider `t('discharging', 'Discharging...')` as a single translatable string.

denniskigen commented on packages/esm-ward-app/src/ward-workspace/patient-discharge/patient-discharge.scss:
```diff
@@ -14,28 +13,18 @@
-
 .workspaceForm {
   margin-top: layout.$spacing-05;
   flex: 1;
```
Nit: double parentheses `((layout.$spacing-05))` — the outer pair is unnecessary.

denniskigen reviewed (dismissed):
Thanks for the iteration here, @sourav-jyoti. The async/await rewrite and test coverage are both solid. A few things to address before this is ready.

sourav-jyoti committed:
changes

sourav-jyoti committed:
Merge branch 'main' into feat/O3-5255

jayasanka-sack committed:
Merge branch 'main' into feat/O3-5255

jayasanka-sack reviewed (dismissed):
@sourav-jyoti the E2E test and build needs fixing. can you take a look and let us know?

sourav-jyoti committed:
changes

sourav-jyoti committed:
changes

sourav-jyoti committed:
changes

sourav-jyoti committed:
changes

sourav-jyoti commented:
@jayasanka-sack done

denniskigen committed:
Merge branch 'main' into feat/O3-5255

denniskigen commented on packages/esm-ward-app/src/ward-workspace/patient-discharge/patient-discharge.workspace.tsx:
```diff
@@ -14,62 +20,79 @@ export default function PatientDischargeWorkspace({
-            title: t('patientWasDischarged', 'Patient was discharged'),
-            kind: 'success',
+      try {
+        const note = dischargeNote;
```
Small polish suggestion: since the discharge note is optional, could we trim it before building `obs`? Otherwise whitespace-only input (e.g. `"   "`) creates a blank note observation instead of behaving like no note.

For example:

```tsx
const note = dischargeNote.trim();
```

It would also be worth adding a small test that whitespace-only input submits with an empty `obs` array.

sourav-jyoti replied:
@denniskigen added .trim()
and also migrated jest to vi

sourav-jyoti replied:
@denniskigen @jayasanka-sack requesting ur review

sourav-jyoti committed:
blank-space

denniskigen committed:
test: fix discharge workspace tests

sourav-jyoti committed:
vite test upgrade

sourav-jyoti committed:
en

sourav-jyoti committed:
en

denniskigen committed:
test: clean up discharge review fixes

chibongho committed:
Merge branch 'main' into feat/O3-5255

denniskigen committed:
Merge branch 'main' into feat/O3-5255

chibongho reviewed (approved):
I discussed this with my team internally. We are ok with this change. (If we need to turn off the confirmation dialog on discharge we'll add a config later.) Thanks!

denniskigen committed:
Align hasUnsavedChanges with onSubmit trim logic

hasUnsavedChanges was using dischargeNote.length > 0, but onSubmit trims
the note before deciding whether to create an obs. A whitespace-only note
would incorrectly trigger the unsaved-changes prompt on close even though
nothing would be persisted. Using trim().length > 0 keeps both consistent.

denniskigen reviewed (approved):
Thanks for seeing this through, @sourav-jyoti — really solid work. The async/await rewrite, error fallback chain, and test coverage are all well done. I pushed one small fix on top:  now uses  to stay consistent with the trim logic in , so whitespace-only input no longer triggers the unsaved-changes prompt.

denniskigen reviewed (approved):
Thanks for seeing this through, @sourav-jyoti. Really solid work. The async/await rewrite, error fallback chain, and test coverage are all well done. I pushed one small fix on top: `hasUnsavedChanges` now uses `dischargeNote.trim().length > 0` to stay consistent with the trim logic in `onSubmit`, so whitespace-only input no longer triggers the unsaved-changes prompt.
